### ISSUE TITLE
mix credo raises (FunctionClauseError) no function clause matching in Regex.match?/2

### ISSUE THREAD

**Issue by sobolevn** [rrrene/credo#215]
### Precheck

* For bugs, please do a quick search and make sure the bug has not yet been reported here

### Environment

* Credo version (`mix credo -v`): 0.4.14
* Erlang/Elixir version (`elixir -v`):
Erlang/OTP 19 [erts-8.0.2] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace] Elixir 1.3.2
* Operating system: OS X 10.11.6

### What where you trying to do?

I was trying to run `mix credo`

### Actual outcome

I have faced some issues with the new release. Here's my configuration:

```elixir
defp deps do
      {:credo, "~> 0.4.14", only: [:dev, :test]},
end
```

Output:

> [11:59:53] sobolev :: MacBook-Pro-Nikita ➜ Documents/github/ecto_autoslug_field ‹master› » mix local.hex
> Found existing archive: /Users/sobolev/.mix/archives/hex-0.12.1.ez.
> Are you sure you want to replace it with "https://repo.hex.pm/installs/1.3.0/hex-0.14.0.ez"? [Yn] y
> 
> creating /Users/sobolev/.mix/archives/hex-0.14.0 [12:06:26] sobolev :: MacBook-Pro-Nikita ➜ Documents/github/ecto_autoslug_field ‹master› » MIX_ENV=test mix deps.get && mix credo --strict Running dependency resolution Dependency resolution completed bunt: 0.1.6 credo: 0.4.14
> Updating credo (Hex package) Checking package (https://repo.hex.pm/tarballs/credo-0.4.14.tar) Fetched package
> ...
> 
> [12:07:08] sobolev :: MacBook-Pro-Nikita ➜ Documents/github/ecto_autoslug_field ‹master› » mix credo 1 ↵
> Checking 4 source files ...
> * (EXIT from #PID<0.70.0>) an exception was raised:
> ** (FunctionClauseError) no function clause matching in Regex.match?/2
> (elixir) lib/regex.ex:159: Regex.match?(~r/(.\w+Controller|.Endpoint|.Mixfile|.Repo|.Router|.\w+Socket|\w+Test|.\w+View)$/, nil)
> (elixir) lib/enum.ex:2511: Enum.do_any?/2
> lib/credo/check/readability/module_doc.ex:42: Credo.Check.Readability.ModuleDoc.traverse/4
> (elixir) lib/macro.ex:296: anonymous fn/4 in Macro.do_traverse_args/4
> (elixir) lib/enum.ex:1247: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
> (elixir) lib/enum.ex:1247: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
> (elixir) lib/macro.ex:262: Macro.do_traverse/4
> (elixir) lib/macro.ex:277: Macro.do_traverse/4
> 
> 12:07:50.528 [error] Task #PID<0.145.0> started from #PID<0.70.0> terminating
> ** (FunctionClauseError) no function clause matching in Regex.match?/2
> (elixir) lib/regex.ex:159: Regex.match?(~r/(.\w+Controller|.Endpoint|.Mixfile|.Repo|.Router|.\w+Socket|\w+Test|.\w+View)$/, nil)
> (elixir) lib/enum.ex:2511: Enum.do_any?/2
> lib/credo/check/readability/module_doc.ex:42: Credo.Check.Readability.ModuleDoc.traverse/4
> (elixir) lib/macro.ex:296: anonymous fn/4 in Macro.do_traverse_args/4
> (elixir) lib/enum.ex:1247: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
> (elixir) lib/enum.ex:1247: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
> (elixir) lib/macro.ex:262: Macro.do_traverse/4
> (elixir) lib/macro.ex:277: Macro.do_traverse/4
> Function: #Function<13.121687013/0 in Credo.Check.Runner.run/2>
> Args: []
> 
> 12:07:50.529 [error] GenServer Credo.Supervisor terminating
> ** (FunctionClauseError) no function clause matching in Regex.match?/2
> (elixir) lib/regex.ex:159: Regex.match?(~r/(.\w+Controller|.Endpoint|.Mixfile|.Repo|.Router|.\w+Socket|\w+Test|.\w+View)$/, nil)
> (elixir) lib/enum.ex:2511: Enum.do_any?/2
> lib/credo/check/readability/module_doc.ex:42: Credo.Check.Readability.ModuleDoc.traverse/4
> (elixir) lib/macro.ex:296: anonymous fn/4 in Macro.do_traverse_args/4
> (elixir) lib/enum.ex:1247: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
> (elixir) lib/enum.ex:1247: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
> (elixir) lib/macro.ex:262: Macro.do_traverse/4
> (elixir) lib/macro.ex:277: Macro.do_traverse/4
> Last message: {:EXIT, #PID<0.70.0>, {:function_clause, [{Regex, :match?, [~r/(.\w+Controller|.Endpoint|.Mixfile|.Repo|.Router|.\w+Socket|\w+Test|.\w+View)$/, nil], [file: 'lib/regex.ex', line: 159]}, {Enum, :do_any?, 2, [file: 'lib/enum.ex', line: 2511]}, {Credo.Check.Readability.ModuleDoc, :traverse, 4, [file: 'lib/credo/check/readability/module_doc.ex', line: 42]}, {Macro, :"-do_traverse_args/4-fun-0-", 4, [file: 'lib/macro.ex', line: 296]}, {Enum, :"-map_reduce/3-lists^mapfoldl/2-0-", 3, [file: 'lib/enum.ex', line: 1247]}, {Enum, :"-map_reduce/3-lists^mapfoldl/2-0-", 3, [file: 'lib/enum.ex', line: 1247]}, {Macro, :do_traverse, 4, [file: 'lib/macro.ex', line: 262]}, {Macro, :do_traverse, 4, [file: 'lib/macro.ex', line: 277]}]}}
> State: {:state, {:local, Credo.Supervisor}, :one_for_one, [{:child, #PID<0.127.0>, Credo.Service.SourceFileIssues, {Credo.Service.SourceFileIssues, :start_link, []}, :permanent, 5000, :worker, [Credo.Service.SourceFileIssues]}, {:child, #PID<0.126.0>, Credo.Service.SourceFileScopes, {Credo.Service.SourceFileScopes, :start_link, []}, :permanent, 5000, :worker, [Credo.Service.SourceFileScopes]}, {:child, #PID<0.125.0>, Credo.Service.SourceFileCodeOnly, {Credo.Service.SourceFileCodeOnly, :start_link, []}, :permanent, 5000, :worker, [Credo.Service.SourceFileCodeOnly]}, {:child, #PID<0.124.0>, Credo.Service.SourceFileWithoutStringAndSigils, {Credo.Service.SourceFileWithoutStringAndSigils, :start_link, []}, :permanent, 5000, :worker, [Credo.Service.SourceFileWithoutStringAndSigils]}], :undefined, 3, 5, [], 0, Supervisor.Default, {:ok, {{:one_for_one, 3, 5}, [{Credo.Service.SourceFileWithoutStringAndSigils, {Credo.Service.SourceFileWithoutStringAndSigils, :start_link, []}, :permanent, 5000, :worker, [Credo.Service.SourceFileWithoutStringAndSigils]}, {Credo.Service.SourceFileCodeOnly, {Credo.Service.SourceFileCodeOnly, :start_link, []}, :permanent, 5000, :worker, [Credo.Service.SourceFileCodeOnly]}, {Credo.Service.SourceFileScopes, {Credo.Service.SourceFileScopes, :start_link, []}, :permanent, 5000, :worker, [Credo.Service.SourceFileScopes]}, {Credo.Service.SourceFileIssues, {Credo.Service.SourceFileIssues, :start_link, []}, :permanent, 5000, :worker, [Credo.Service.SourceFileIssues]}]}}}

Here's the full source code: https://github.com/sobolevn/ecto_autoslug_field

This issue was originally posted in https://github.com/rrrene/credo/issues/210

**Comment by sstern6** [comment]
Thanks for the issue @sobolevn.

Can you please do me a favor and update credo to `0.5.1` per the latest version on hex: https://hex.pm/packages/credo and try rerunning just in case?

Also, I believe I know what the problem is and should be fixed by this pr, https://github.com/rrrene/credo/pull/204.  Might need to wait for the next minor release.

It seems like in your project, you are running credo against nested `defmodules` in some files which was mentioned in this issue https://github.com/rrrene/credo/issues/201.  Ill look deeper into this to confirm.

cc @rrrene

**Comment by rrrene** [comment]
Correct. Credo on `master` does not longer fail on `ecto_autoslug_field` :+1:

**Comment by sobolevn** [comment]
Sorry it took me long. Confirming, that it is working with `master`. But it fails with `~ 0.5`. Logs:

> [19:48:27] sobolev :: MacBook-Pro-Nikita  ➜  Documents/github/ecto_autoslug_field ‹master› » mix deps.get
> Running dependency resolution
> Dependency resolution completed
>   bunt: 0.1.6
>   credo: 0.5.1
> * Updating credo (Hex package)
>   Checking package (https://repo.hex.pm/tarballs/credo-0.5.1.tar)
>   Fetched package
> [19:48:44] sobolev :: MacBook-Pro-Nikita  ➜  Documents/github/ecto_autoslug_field ‹master*› » mix credo
> ==> credo
> Compiling 120 files (.ex)
> Generated credo app
> Checking 4 source files ...
> ** (EXIT from #PID<0.70.0>) an exception was raised:
>     ** (FunctionClauseError) no function clause matching in Regex.match?/2
>         (elixir) lib/regex.ex:159: Regex.match?(~r/(\.\w+Controller|\.Endpoint|\.Mixfile|\.Repo|\.Router|\.\w+Socket|\w+Test|\.\w+View)$/, nil)
>         (elixir) lib/enum.ex:2511: Enum.do_any?/2
>         lib/credo/check/readability/module_doc.ex:42: Credo.Check.Readability.ModuleDoc.traverse/4
>         (elixir) lib/macro.ex:296: anonymous fn/4 in Macro.do_traverse_args/4
>         (elixir) lib/enum.ex:1247: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
>         (elixir) lib/enum.ex:1247: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
>         (elixir) lib/macro.ex:262: Macro.do_traverse/4
>         (elixir) lib/macro.ex:277: Macro.do_traverse/4
> 
> 19:48:54.150 [error] Task #PID<0.891.0> started from #PID<0.70.0> terminating
> ** (FunctionClauseError) no function clause matching in Regex.match?/2
>     (elixir) lib/regex.ex:159: Regex.match?(~r/(\.\w+Controller|\.Endpoint|\.Mixfile|\.Repo|\.Router|\.\w+Socket|\w+Test|\.\w+View)$/, nil)
>     (elixir) lib/enum.ex:2511: Enum.do_any?/2
>     lib/credo/check/readability/module_doc.ex:42: Credo.Check.Readability.ModuleDoc.traverse/4
>     (elixir) lib/macro.ex:296: anonymous fn/4 in Macro.do_traverse_args/4
>     (elixir) lib/enum.ex:1247: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
>     (elixir) lib/enum.ex:1247: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
>     (elixir) lib/macro.ex:262: Macro.do_traverse/4
>     (elixir) lib/macro.ex:277: Macro.do_traverse/4
> Function: #Function<13.121687013/0 in Credo.Check.Runner.run/2>
>     Args: []
> 
> 19:48:54.151 [error] GenServer Credo.Supervisor terminating
> ** (FunctionClauseError) no function clause matching in Regex.match?/2
>     (elixir) lib/regex.ex:159: Regex.match?(~r/(\.\w+Controller|\.Endpoint|\.Mixfile|\.Repo|\.Router|\.\w+Socket|\w+Test|\.\w+View)$/, nil)
> ...
> [19:48:54] sobolev :: MacBook-Pro-Nikita  ➜  Documents/github/ecto_autoslug_field ‹master*› » mix deps.get                                                  1 ↵
> * Getting credo (https://github.com/rrrene/credo.git)
> Cloning into '/Users/sobolev/Documents/github/ecto_autoslug_field/deps/credo'...
> remote: Counting objects: 4047, done.        
> remote: Compressing objects: 100% (140/140), done.        
> remote: Total 4047 (delta 78), reused 0 (delta 0), pack-reused 3907        
> Receiving objects: 100% (4047/4047), 1.05 MiB | 367.00 KiB/s, done.
> Resolving deltas: 100% (2730/2730), done.
> Checking connectivity... done.
> Running dependency resolution
> Dependency resolution completed
>   bunt: 0.1.6
> [19:53:44] sobolev :: MacBook-Pro-Nikita  ➜  Documents/github/ecto_autoslug_field ‹master*› » mix credo   
> ==> credo
> Compiling 122 files (.ex)
> Generated credo app
> Checking 4 source files ...
> 
> Please report incorrect results: https://github.com/rrrene/credo/issues
> 
> Analysis took 0.1 seconds (0.00s to load, 0.1s running checks)
> 29 mods/funs, found no issues.
> 
> Use `--strict` to show all issues, `--help` for options.

**Comment by rrrene** [comment]
Perfect. I will do my best to find time tonight (to release a new version). Thx for excellent bug report! It helped a lot :+1:

**Comment by sstern6** [comment]
@rrrene saw you bumped to 0.5.2, @sobolevn can you check this works on latest release so we can either close this issue or troubleshoot?

Thanks everyone!

**Comment by sobolevn** [comment]
You are officially awesome! 

> [13:07:04] sobolev :: MacBook-Pro-Nikita  ➜  Documents/github/ecto_autoslug_field ‹master*› » mix deps.get
> Running dependency resolution
> Dependency resolution completed
>   credo: 0.5.2
> * Updating credo (Hex package)
>   Checking package (https://repo.hex.pm/tarballs/credo-0.5.2.tar)
>   Fetched package
> [13:07:24] sobolev :: MacBook-Pro-Nikita  ➜  Documents/github/ecto_autoslug_field ‹master*› » mix credo 
> ==> credo
> Compiling 122 files (.ex)
> Generated credo app
> Checking 4 source files ...
> 
> Please report incorrect results: https://github.com/rrrene/credo/issues
> 
> Analysis took 0.1 seconds (0.00s to load, 0.1s running checks)
> 29 mods/funs, found no issues.
> 
> Use `--strict` to show all issues, `--help` for options.

**Comment by rrrene** [comment]
@sobolevn I will link to this comment whenever someone disputes my official awesomeness in the future 😁  :+1:
